### PR TITLE
LightningModule support for Lite

### PR DESCRIPTION
**Pull request by lightningforever**
## What does this PR do?

Adds a LightningModule integration for Lite.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/core/module.py [hunk 1]
```diff
@@ -118,6 +122,10 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self._register_sharded_tensor_state_dict_hooks_if_available()
         self._compiler_ctx: Optional[Dict[str, Any]] = None
 
+        # attributes only used when using fabric
+        self._fabric: Optional["lf.Fabric"] = None
+        self._optimizers: List[_FabricOptimizer] = []
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:127]: I would make it clear that this is fabric specific

```suggestion
        self._fabric_optimizers: List[_FabricOptimizer] = []
```

File: src/pytorch_lightning/core/module.py [hunk 2]
```diff
@@ -177,6 +191,8 @@ def lr_schedulers(self) -> Union[None, List[LRSchedulerPLType], LRSchedulerPLTyp
     def trainer(self) -> "pl.Trainer":
         if not self._jit_is_scripting and self._trainer is None:
             raise RuntimeError(f"{self.__class__.__qualname__} is not attached to a `Trainer`.")
+        if self._fabric is not None:
+            return _TrainerFabricShim(self._fabric)  # type: ignore[return-value]
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:195]: Shouldn't this be above the RuntimeError?

File: src/pytorch_lightning/core/module.py [hunk 3]
```diff
@@ -188,6 +204,21 @@ def trainer(self, trainer: Optional["pl.Trainer"]) -> None:
             trainer = weakref.proxy(trainer)
         self._trainer = trainer
 
+    @property
+    def fabric(self) -> "lf.Fabric":
+        if not self._jit_is_scripting and self._fabric is None:
+            raise RuntimeError(f"{self.__class__.__qualname__} is not attached to a `Fabric`.")
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:210]: Branding:

```suggestion
            raise RuntimeError(f"{self.__class__.__qualname__} is not attached to `Fabric`.")
```

File: src/pytorch_lightning/core/module.py [hunk 4]
```diff
@@ -1429,7 +1466,8 @@ def toggle_optimizer(self, optimizer: Union[Optimizer, LightningOptimizer], opti
         # Iterate over all optimizer parameters to preserve their `requires_grad` information
         # in case these are pre-defined during `configure_optimizers`
         param_requires_grad_state = {}
-        for opt in self.trainer.optimizers:
+        optimizers = self._optimizers if self._fabric is not None else self.trainer.optimizers
+        for opt in optimizers:
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:1470]: Couldn't we do this via the shim? Is `Fabric.optimizers` not useful?

Even if we don't need to hold an owning reference, it could be a weak reference.

**Review by carmocca — COMMENTED** [review]
Will approve when it's not a draft! Honestly this was less code than I was expecting

File: src/pytorch_lightning/core/module.py (hunk 4 above)
Comment (lightningforever) [reply src/pytorch_lightning/core/module.py]: > Is Fabric.optimizers not useful?

I want to avoid keeping track of modules and optimizers inside Fabric. Fabric has a very "functional" approach and this makes it flexible, and at the same time easy to understand. Unless it can improve the UX, I vote for not doing that.

Comment (lantiga) [reply src/pytorch_lightning/core/module.py]: I agree, let's not do that for now, we can always change our minds later but let's keep it leaner for now.

**Review by lantiga — APPROVED** [review]
Love this ❤️!!!

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/core/module.py [hunk 5]
```diff
@@ -189,6 +206,19 @@ def trainer(self, trainer: Optional["pl.Trainer"]) -> None:
...
         self._trainer = trainer
 
+    @property
+    def fabric(self) -> Optional["lf.Fabric"]:
+        return self._fabric
+
+    @fabric.setter
+    def fabric(self, fabric: Optional["lf.Fabric"]) -> None:
+        for v in self.children():
+            if isinstance(v, LightningModule):
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:216]: Isn't this a bug?

```suggestion
            if isinstance(v, lf.Fabric):
```
